### ISSUE TITLE
Feature Idea: Checkpoint API "split" assert_consumed

### ISSUE THREAD

**Issue by sleighsoft** [tensorflow/tensorflow#21156]
I would like to see a function that is called something like `not_consumed()` for the checkpoint API here (https://www.tensorflow.org/api_docs/python/tf/train/Checkpoint#restore) that returns all objects in the dependency graph that were not found in the checkpoint but without raising a python assertion like `assert_consumed` does.

I would use it because sometimes I want to make sure that I restored certain variables but about others I do not care if they are restored or not.

In general I believe this would be a minor change to the code (e.g. splitting `assert_consumed` into two) that would greatly help to write tests to validate that variables are indeed restored and probably also be helpful in other scenarios as well.

What is your take on this @allenlavoie ?

**Comment by allenlavoie** [comment]
Seems reasonable to me. The assertion as-is is quite strict.

If it's part of the public API, I'd be worried about returning random proto nonsense for things in the checkpoint that didn't get loaded. But maybe something like `variables_not_restored` for Python variable objects which don't have a checkpointed value? It's still slightly odd, since the collection will grow over time as graph building (or eager execution) progresses and variables get created. But seems useful enough.

**Comment by sleighsoft** [comment]
Yes, something along these lines. If possible I think that returning something like a tuple `variables_not_restored=(variables_not_in_checkpointing, variables_in_checkpoint_but_not_in_current_dependency_graph)`... a little verbose here :D

**Comment by allenlavoie** [comment]
Not at the moment. I have changes pending to configure garbage collection of checkpoints (will probably be a separate object), and for other ways of numbering checkpoints. Did you have something else in mind?

**Comment by sleighsoft** [comment]
Currently I would just like to set the `max_to_keep` setting of `tf.train.Saver` to a different value.

**Comment by allenlavoie** [comment]
@sleighsoft max_to_keep should now be addressed with https://github.com/tensorflow/tensorflow/commit/7a81491366afb444efb914b93594a11fdfc19896

Usage example from the docstring:
```
import tensorflow as tf
checkpoint = tf.train.Checkpoint(optimizer=optimizer, model=model)
manager = tf.contrib.checkpoint.CheckpointManager(
    checkpoint, directory="/tmp/model", max_to_keep=5)
status = checkpoint.restore(manager.latest_checkpoint)
while True:
  # train
  manager.save()
 ```

**Comment by sleighsoft** [comment]
The CheckpointManager looks great :)
Exactly what I imagined!

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @allenlavoie: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @allenlavoie: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @allenlavoie: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @allenlavoie: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @allenlavoie: It has been 77 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rmothukuru** [comment]
@sleighsoft,
Can you please confirm if this issue is resolved and if we can close it? Thanks!

**Comment by sleighsoft** [comment]
I guess @allenlavoie is the better person to ask here. I cannot answer as I haven't had much exposure recently

**Comment by rmothukuru** [comment]
Closing the issue as the **`API`**, [CheckPointManager](https://www.tensorflow.org/api_docs/python/tf/train/CheckpointManager) addresses the issue that has been raised. Thanks!
